This PR adds the class 'DQ_SerialVrepRobot'

ffasilva opened the pull request:
# Main instructions

By submitting this pull request, you automatically agree that you have read and accepted the following conditions:

- Anyone wanting to propose a new modification or introduce new functionality should reach out to the team first, as proposed modifications that do not comply with the library's development philosophy and style, do not follow the library's architecture, do not introduce a clear and general benefit to the library other than to the person who proposed the modification will likely be rejected with no further discussion.
- Support for DQ Robotics is given voluntarily and it is not the developers' role to educate and/or convince anyone of their vision.
- Any possible response and its timeliness will be based on the relevance, accuracy, and politeness of a request and the following discussion.
- You are familiar with the [development workflow](https://github.com/dqrobotics/matlab/blob/master/CONTRIBUTING.md).
- Each pull request should contain only individual changes (several changes of the same type **are allowed** on the same pull request).
- Refactoring or modifying internal implementation that is working is not allowed unless comprehensively discussed with and approved by @bvadorno and @mmmarinho.


## Description of changes
Hi, @dqrobotics/developers,

I've added the the class `DQ_SerialVrepRobot`. It follows the implementation of its [C++ counterpart](https://github.com/dqrobotics/cpp-interface-vrep/blob/master/src/dqrobotics/interfaces/vrep/DQ_VrepRobot.cpp).

I also had to update the classes `DQ_VrepRobot`, `LBR4pVrepRobot`, and `YouBotVrepRobot` for compatibility with the new signatures imposed by `DQ_SerialVrepRobot`. 

An example of use is given in [9](https://github.com/dqrobotics/matlab-examples/pull/9).

## Rationale
The class `DQ_SerialVrepRobot` already exists in the C++ library but was missing on MATLAB.

ffasilva commented:
Hi @dqrobotics/developers,

Could we please prioritize this pull request? I cannot do the fourth, and last, lesson of the [learning-dqrobotics-with-coppeliasim](https://github.com/dqrobotics/learning-dqrobotics-with-coppeliasim) tutorials without it being merged into the master.

The only [robots available](https://github.com/dqrobotics/matlab/tree/master/interfaces/vrep/robots) in the public DQ Robotics interface with CoppeliaSim are [LBR4pVrepRobot](https://github.com/dqrobotics/matlab/blob/master/interfaces/vrep/robots/LBR4pVrepRobot.m) and [YouBotVrepRobot](https://github.com/dqrobotics/matlab/blob/master/interfaces/vrep/robots/YouBotVrepRobot.m). Besides having the obsolete methods of `get_q_from_vrep` and `send_q_to_vrep` (which I'll have to update in the [third lesson](https://github.com/dqrobotics/learning-dqrobotics-with-coppeliasim/pull/4) of the tutorial), they only work for scenes with dynamics disabled because they use the `vrep_interface.set_joint_positions` method for actuating on the robots.

The fourth lesson of the tutorial is about dynamic simulations and methods such as `set_joint_target_positions`, `get_joint_velocities`, etc. I could do some weird workaround of submitting a new pull request just updating `LBR4pVrepRobot` and `YouBotVrepRobot` to continue with the tutorials, but this fix is already covered by [ace84af](https://github.com/dqrobotics/matlab/pull/108/commits/ace84afa2b2bc734d820351a3f138049200595d7) in this pull request.

Kind regards,
Frederico

bvadorno commented:
Hi @mmmarinho,

Can you please take this one? As you designed the original classes, you are better positioned to review the new classes.

@ffasilva, could you please clarify if this modification will break the second lesson of the CoppeliaSim-with-DQ-Robotics tutorial? 

Kind regards,
Bruno

mmmarinho commented:
@bvadorno sure!

@juanjqo are you taking a look at this one first or should I?

ffasilva commented:
Hi @bvadorno,

> @ffasilva, could you please clarify if this modification will break the second lesson of the CoppeliaSim-with-DQ-Robotics tutorial?

I don't see how it would. [Lesson 2](https://github.com/dqrobotics/learning-dqrobotics-with-coppeliasim/pull/3) doesn't use any robotic manipulator. However, this pull request will break the ending of lesson 3, hence why I left it as a [draft](https://github.com/dqrobotics/learning-dqrobotics-with-coppeliasim/pull/4).

Kind regards,
Frederico

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+%           >> vi.connect('127.0.0.1',19997);
+%           >> vrep_robot = DQ_SerialVrepRobot("my_robot", 7, "my_robot", vi);
+%           >> vi.start_simulation();
+%           >> vrep_robot.get_q_from_vrep();
```
This is probably the old usage `get_q_from_vrep`, better to update

ffasilva replied:
Updated in [0fd50ac](https://github.com/dqrobotics/matlab/pull/108/commits/0fd50ac04e454399a7b2d588758ee59b7cb2e533).

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+%   DQ_SerialVrepRobot Methods (Protected):
+%       update_manipulator_dynamic_parameters - Updates the dynamic parameters of the serial robot in the CoppeliaSim scene
+
+% (C) Copyright 2018-2023 DQ Robotics Developers
```
-2024

ffasilva replied:
Updated in [ef38693](https://github.com/dqrobotics/matlab/pull/108/commits/ef3869349e089c0eed2b1147cb312e60bb413e87).

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+%
+% DQ Robotics website: dqrobotics.github.io
+%
+% Contributors to this file:
```
Please add reference to the original C++ file as the source for this.

ffasilva replied:
Add in [217f7cd](https://github.com/dqrobotics/matlab/pull/108/commits/217f7cd0eb805b6b3cb79756f271f0f5979775f8).

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+        end
+
+        function set_configuration_space_positions(obj, q)
+            % This method sets the joint configurations to the robot in the CoppeliaSim scene.
```
"of the robot", to properly link with "set". This happens in the other `set` methods.

ffasilva replied:
Fixed in [aa73594](https://github.com/dqrobotics/matlab/pull/108/commits/aa73594843477cc3a31225e2eea32d659d33df2f).

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+            %     vrep_robot.set_configuration_space_positions(q);
+            %     
+            %     Note that this calls "set_joint_positions" in the remoteAPI, meaning that it
+            %     is only suitable for passive joints.
```
Please double check if the "passive" and "active" joint terminology is what is being used in CoppeliaSim.

ffasilva replied:
Updated in [3c4d66e](https://github.com/dqrobotics/matlab/pull/108/commits/3c4d66e9f1c4d12fd2d9ec61fb1274f7c845fed6). It should probably be changed in the [C++ version](https://github.com/dqrobotics/cpp-interface-vrep/blob/master/src/dqrobotics/interfaces/vrep/DQ_SerialVrepRobot.cpp) as well.

mmmarinho commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,240 @@
+            % This method sets the joint velocities of the robot in the CoppeliaSim scene as a target velocity for the joint controllers.
+            % Usage:
+            %     set_target_configuration_space_positions(q);  
+            %          v_target: The target joint velocities of the robot in the CoppeliaSim scene.
```
I'm a bit confused about the `v_target` placement. Shouldn't `v_target` be mentioned the usage?
This happens in other methods as well.

ffasilva replied:
It was copy and paste gore. Fixed in [e0f1a98](https://github.com/dqrobotics/matlab/pull/108/commits/e0f1a983b8813bc7a153e034f163b04f0c997ff5).

mmmarinho commented on interfaces/vrep/DQ_VrepRobot.m:
```diff
@@ -30,7 +30,16 @@
+%     1. Murilo Marques Marinho - murilo@nml.t.u-tokyo.ac.jp
+%       - Responsible for the original implementation.
+%     2. Frederico Fernandes Afonso Silva (frederico.silva@ieee.org)
+%       - Altered the names of the following methods to ensure
```
Careful with the backwards compatibility. It is better to add an alias for the method than to rename it.

ffasilva replied:
Backwards compatibility added in [a02ad1b](https://github.com/dqrobotics/matlab/pull/108/commits/a02ad1ba01016269077e04b6465c43d803049bd1).

mmmarinho commented on interfaces/vrep/robots/YouBotVrepRobot.m:
```diff
@@ -17,8 +17,8 @@
 %       will become "youBot#0", a third robot, "youBot#1", and so on.
 %
 %   YouBotVrepRobot Methods:
-%       send_q_to_vrep - Sends the joint configurations to VREP
```
Careful removing methods, we need at least one version of backwards compatibility

ffasilva replied:
Backwards compatibility added in [a02ad1b](https://github.com/dqrobotics/matlab/pull/108/commits/a02ad1ba01016269077e04b6465c43d803049bd1).

mmmarinho replied:
In the review view, this is still deleted and not added anywhere else. Maybe I'm doing something wrong?

ffasilva replied:
In [a02ad1b](https://github.com/dqrobotics/matlab/pull/108/commits/a02ad1ba01016269077e04b6465c43d803049bd1) I added backwards compatibility to both `send_q_to_vrep` and `get_q_from_vrep` to the class `DQ_VrepRobot`, where they used to be abstract methods.

``` matlab
function send_q_to_vrep(obj, q)
   % For backwards compatibility only. Do not use this method.

   warning('Deprecated. Use set_configuration_space_positions() instead.')
   obj.set_configuration_space_positions(q)
end

function q = get_q_from_vrep(obj)
   % For backwards compatibility only. Do not use this method.

   warning('Deprecated. Use get_configuration_space_positions() instead.')
   q = obj.get_configuration_space_positions();
end
```

As such, the use of something similar to

``` matlab
vi = DQ_VrepInterface();
vi.connect('127.0.0.1',19997);
vrep_robot = YouBotVrepRobot("youBot", vi);
vrep_robot.get_q_from_vrep
```

yields the result
``` matlab
Warning: Deprecated. Use get_configuration_space_positions() instead. 
> In DQ_VrepRobot/get_q_from_vrep (line 65) 

ans =

    1.2068
   -1.8492
    3.6463
    0.0001
    0.5395
    0.9149
    1.2685
   -0.0001
```

That's the desired result, right? The deprecated method `send_q_to_vrep` works similarly.

mmmarinho replied:
Hi, @ffasilva ,

The desired result is to leave the original code unchanged as much as possible and work the new code around it. 

I understand why you chose to do it this way, but I'll condition the acceptance of this to these examples working without modification:
https://github.com/dqrobotics/matlab-examples/tree/master/vrep

ffasilva replied:
Hi @mmmarinho,

I see what you mean, but I figured that deprecating the methods in all the concrete classes that inherit from `DQ_VrepRobot`, rather than just doing it directly on the abstract class itself, would have generated even more modifications to the original code.

As for your request, the deprecation of the methods is working as intended. The [simulation-ram-paper](https://github.com/dqrobotics/matlab-examples/tree/master/vrep/simulation-ram-paper) runs without any problems:

![simulation_ram_paper](https://github.com/user-attachments/assets/51f4c3b5-d028-4a13-8eff-78716e80f95a)

However, the [constrained_kinematic_control](https://github.com/dqrobotics/matlab-examples/tree/master/vrep/constrained_kinematic_control) example don't use the VrepRobot classes at all. Which means it obviously works, but it also doesn't help us validate my modifications. I raised an [issue](https://github.com/dqrobotics/matlab-examples/issues/11) to update this example in the future.

Kind regards,
Frederico

bvadorno replied:
Thanks, @ffasilva. Your reasoning is sensible as you have minimized the number of modifications, and the example still runs. Therefore, your solution is acceptable.

mmmarinho reviewed (changes requested):
@ffasilva 

Thanks for this. I've written a few comments about the suggested changes.

ffasilva commented:
Hi @mmmarinho,

Thank you for the review. I implemented the requested modifications and fixed small details in the documentation. Please let me know if there is still something missing.

Kind regards,
Frederico

mmmarinho commented on interfaces/vrep/robots/YouBotVrepRobot.m:
```diff
@@ -41,28 +41,21 @@
+        function obj = YouBotVrepRobot(robot_name, vrep_interface)
+            obj@DQ_SerialVrepRobot("youBot", 7, robot_name, vrep_interface);
             
+            %% youBot don't follow the standard name convention on CoppeliaSim. Also, the use of 'set_names()', as is done in the C++ implementation, is not supported on a constructor in MATLAB
```
There are some ways to fix this but how about

> youBot` **does not** follow the standard **naming** convention **of/in** CoppeliaSim.

ffasilva replied:
Fixed in [fa35f1b](https://github.com/dqrobotics/matlab/pull/108/commits/fa35f1bd334ede9c8bd6aadb1fcc0d504edc1786).

mmmarinho reviewed (changes requested):
Hi @ffasilva. Thank you, I've added some comments.

bvadorno commented on interfaces/vrep/DQ_VrepRobot.m:
```diff
@@ -4,13 +4,11 @@
+%       get_configuration_space_positions - Obtains the joint configurations from VREP
 
-% (C) Copyright 2020 DQ Robotics Developers
+% (C) Copyright 2018-2024 DQ Robotics Developers
```
DQ Robotics, as we know it now, started in [2011](https://github.com/dqrobotics/matlab/commit/8ccc2487e6d45d37162f9daaf7ad60a9c5f1ae33). Please amend the copyright date (remember that this is legally binding, so we must take it seriously.)

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%           >> vi.connect('127.0.0.1',19997);
+%           >> vrep_robot = DQ_SerialVrepRobot("my_robot", 7, "my_robot", vi);
+%           >> vi.start_simulation();
+%           >> vrep_robot.get_configuration_space_positions();
```
Let's change it to `get_configuration()` and deprecate `get_configuration_space_positions()`.

ffasilva replied:
Updated comment in [c934c9c](https://github.com/dqrobotics/matlab/pull/108/commits/c934c9cca0aed4d1327b7c9c18b07047b580702a). However, there is no need to deprecate `get_configuration_space_positions()` as this is the pull requesting proposing the first MATLAB version of `DQ_SerialVrepRobot`. We willl have to deprecate the methods in the [C++ version](https://github.com/dqrobotics/cpp-interface-vrep/blob/master/include/dqrobotics/interfaces/vrep/DQ_SerialVrepRobot.h), though.

mmmarinho replied:
@ffasilva nothing only the `master` branch needs to be deprecated. Only what ended up in a release version.

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%
+%   DQ_SerialVrepRobot Methods:
+%       get_joint_names - Gets the joint names of the robot in the CoppeliaSim scene.
+%       set_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene.
```
Let's change it to `get_configuration()` and deprecate `get_configuration_space_positions()`.

ffasilva replied:
Comment updated in [bee94d7](https://github.com/dqrobotics/matlab/pull/108/commits/bee94d72953ad73df020f6e2cec14cbb5b9542f2).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%
+%   DQ_SerialVrepRobot Methods:
+%       get_joint_names - Gets the joint names of the robot in the CoppeliaSim scene.
+%       set_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene.
```
Let's change it to `set_configuration()` and deprecate `set_configuration_space_positions()`.

ffasilva replied:
Updated comment in [c9ef19e](https://github.com/dqrobotics/matlab/pull/108/commits/c9ef19e7688698726c03b5f13f5deb7aa16abcdc).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%       get_joint_names - Gets the joint names of the robot in the CoppeliaSim scene.
+%       set_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene.
+%       get_configuration_space_positions - Gets the joint configurations of the robot in the CoppeliaSim scene.
+%       set_target_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene as a target configuration for the joint controllers.
```
`set_target_configuration()`

ffasilva replied:
Updated in [5099cb7](https://github.com/dqrobotics/matlab/pull/108/commits/5099cb7990f77938d21c512801e68df89673e48f).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%       set_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene.
+%       get_configuration_space_positions - Gets the joint configurations of the robot in the CoppeliaSim scene.
+%       set_target_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene as a target configuration for the joint controllers.
+%       get_configuration_space_velocities - Gets the joint velocities of the robot in the CoppeliaSim scene.
```
`get_configuration_velocities()`

ffasilva replied:
Updated in [c3356ae](https://github.com/dqrobotics/matlab/pull/108/commits/c3356aeb51058ab0faf8ce7a60f92f8d4c711c6a).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%       get_configuration_space_positions - Gets the joint configurations of the robot in the CoppeliaSim scene.
+%       set_target_configuration_space_positions - Sets the joint configurations of the robot in the CoppeliaSim scene as a target configuration for the joint controllers.
+%       get_configuration_space_velocities - Gets the joint velocities of the robot in the CoppeliaSim scene.
+%       set_target_configuration_space_velocities - Sets the joint velocities of the robot in the CoppeliaSim scene as a target velocity for the joint controllers.
```
`set_target_configuration_velocities()`

ffasilva replied:
Updated in [6ac8182](https://github.com/dqrobotics/matlab/pull/108/commits/6ac8182b6c2686c95e05a1ad3b10b2878b3dfbaa).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+%   DQ_SerialVrepRobot Methods (Protected):
+%       update_manipulator_dynamic_parameters - Updates the dynamic parameters of the serial robot in the CoppeliaSim scene
+
+% (C) Copyright 2018-2024 DQ Robotics Developers
```
2011-2024

ffasilva replied:
Done in [107be2c](https://github.com/dqrobotics/matlab/pull/108/commits/107be2c48fde8a3670b5689d865f4825ae9929dc).

bvadorno commented on interfaces/vrep/DQ_SerialVrepRobot.m:
```diff
@@ -0,0 +1,242 @@
+            joint_names = obj.joint_names;
+        end
+
+        function set_configuration_space_positions(obj, q)
```
Please change those methods according to my comments above.

ffasilva replied:
Names changed in [8a60f03](https://github.com/dqrobotics/matlab/pull/108/commits/8a60f03ab59b96016931954e1f8b119ee9bf1843).

bvadorno commented on interfaces/vrep/DQ_VrepRobot.m:
```diff
@@ -4,13 +4,11 @@
-%       robot
+%   DQ_VrepRobot Methods (Abstract):
+%       set_configuration_space_positions - Sends the joint configurations to VREP
+%       get_configuration_space_positions - Obtains the joint configurations from VREP
```
Please change the methods' names according to my comments on the `DQ_SerialVrepRobot.m` file.

ffasilva replied:
Names changed in [0cf51a6](https://github.com/dqrobotics/matlab/pull/108/commits/0cf51a6b35c6a49f577d5aa3494ed10e7aa136f0).

bvadorno commented on interfaces/vrep/robots/LBR4pVrepRobot.m:
```diff
@@ -17,11 +17,9 @@
 %       kinematics - Obtains the DQ_Kinematics implementation of this robot
 
-% (C) Copyright 2020 DQ Robotics Developers
+% (C) Copyright 2018-2024 DQ Robotics Developers
```
2011-2024

ffasilva replied:
[17f4506](https://github.com/dqrobotics/matlab/pull/108/commits/17f45069b45f793c0e0a2408367c81a2723c54e8).

bvadorno commented on interfaces/vrep/robots/LBR4pVrepRobot.m:
```diff
@@ -8,7 +8,7 @@
 %           >> vrep_robot = LBR4pVrepRobot("LBR4p", vi);
 %           >> vi.start_simulation();
-%           >> robot.get_q_from_vrep();
+%           >> robot.get_configuration_space_positions();
```
Please change the methods' names according to my comments on the other files. The legacy names should be deprecated.

ffasilva replied:
Updated in [ecab5b5](https://github.com/dqrobotics/matlab/pull/108/commits/ecab5b5d6d0eacbe3a37633a3e48c4f877974075).

bvadorno commented on interfaces/vrep/robots/YouBotVrepRobot.m:
```diff
@@ -8,7 +8,7 @@
 %           >> vrep_robot = YouBotVrepRobot("youBot", vi);
 %           >> vi.start_simulation();
-%           >> robot.get_q_from_vrep();
+%           >> robot.get_configuration_space_positions();
```
Same as in the other files.

ffasilva replied:
Updated in [05873d5](https://github.com/dqrobotics/matlab/pull/108/commits/05873d58d643581beeb2140712c1e05b56cdc928).

bvadorno commented on interfaces/vrep/robots/YouBotVrepRobot.m:
```diff
@@ -17,11 +17,11 @@
 %   YouBotVrepRobot Methods:
-%       send_q_to_vrep - Sends the joint configurations to VREP
-%       get_q_from_vrep - Obtains the joint configurations from VREP
+%       set_configuration_space_positions - Sends the joint configurations to VREP
```
Same as before: please change the name and deprecate legacy names.

ffasilva replied:
Updated in [8b53f6d](https://github.com/dqrobotics/matlab/pull/108/commits/8b53f6ddc4ffc68d5f36e70b509d0ff269dd3330).

bvadorno commented on interfaces/vrep/robots/YouBotVrepRobot.m:
```diff
@@ -17,11 +17,11 @@
 %       kinematics - Obtains the DQ_Kinematics implementation of this robot
 
-% (C) Copyright 2020 DQ Robotics Developers
+% (C) Copyright 2018-2024 DQ Robotics Developers
```
2011-2024

ffasilva replied:
Updated in [badb11a](https://github.com/dqrobotics/matlab/pull/108/commits/badb11aa7d5893182e154ae0c97da61740393e5b).

bvadorno reviewed (changes requested):
Please implement the requested changes.

bvadorno reviewed (approved):
Hi @ffasilva,

Please create a pull request for the C++ version to ensure compatibility with the new names.

Thanks!
Bruno
